### PR TITLE
use cupla instead of cuda prefix for function calls

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Avoid that PMacc/PIConGPU depends on the preprocessor renaming to use
cupla.

Since PMacc is still using sometimes native cuda it is hard to see where cupla and whee CUDA is used. I also work on making the precompiler stuff in cupla optional to increase the compatibility of cupla.

### CODE REVIEW

File: include/pmacc/cuSTL/container/copier/Memcopy.hpp [hunk 1]
```diff
@@ -48,9 +48,9 @@ struct Memcopy<1>
                     const Type* source, const math::Size_t<0>, const math::Size_t<1>& size,
                     flags::Memcopy::Direction direction)
     {
-            const cudaMemcpyKind kind[] = {cudaMemcpyHostToDevice, cudaMemcpyDeviceToHost,
-                                     cudaMemcpyHostToHost, cudaMemcpyDeviceToDevice};
-            CUDA_CHECK(cudaMemcpy(dest, source, sizeof(Type) * size.x(), kind[direction]));
+            const cuplaMemcpyKind kind[] = {cuplaMemcpyHostToDevice, cuplaMemcpyDeviceToHost,
+                                     cuplaMemcpyHostToHost, cuplaMemcpyDeviceToDevice};
+            CUDA_CHECK(cuplaMemcpy(dest, source, sizeof(Type) * size.x(), kind[direction]));
```
Comment (sbastrakov) [inline include/pmacc/cuSTL/container/copier/Memcopy.hpp]: Should this still be `CUDA_CHECK` ? (same in other places)

File: include/pmacc/cuplaHelper/ValidateCall.hpp [hunk 2]
```diff
@@ -31,27 +31,27 @@ namespace pmacc
...
 
 /**
- * Print a cuda error message including file/line info to stderr and raises an exception
+ * Print a cupla error message including file/line info to stderr and raises an exception
  */
-#define PMACC_PRINT_CUDA_ERROR_AND_THROW(cudaError, msg) \
+#define PMACC_PRINT_CUDA_ERROR_AND_THROW(cuplaError, msg) \
     PMACC_PRINT_CUDA_ERROR(msg);                         \
-    throw std::runtime_error(std::string("[CUDA] Error: ") + std::string(cudaGetErrorString(cudaError)))
+    throw std::runtime_error(std::string("[CUDA] Error: ") + std::string(cuplaGetErrorString(cuplaError)))
```
Comment (sbastrakov) [inline include/pmacc/cuplaHelper/ValidateCall.hpp]: Perhaps rename this and the following macro with CUDA as well? Or at least renaming in the output `[CUDA] Error` to `[Cupla] Error` makes sense to me.

File: include/pmacc/dimensions/DataSpace.hpp [hunk 3]
```diff
@@ -60,7 +60,7 @@ namespace pmacc
          * constructor.
          * Sets size of all dimensions from cuda dim3.
          */
-        HDINLINE explicit DataSpace(dim3 value)
+        HDINLINE DataSpace(cupla::dim3 value)
```
Comment (sbastrakov) [inline include/pmacc/dimensions/DataSpace.hpp]: Are these `explicit` changes related to the PR?

File: include/pmacc/math/vector/Vector.hpp [hunk 4]
```diff
@@ -525,9 +525,9 @@ struct Vector : private T_Storage<T_Type, T_dim>, protected T_Accessor, protecte
         return stream.str();
     }
 
-    HDINLINE dim3 toDim3() const
+    HDINLINE cupla::dim3 toDim3() const
```
Comment (sbastrakov) [inline include/pmacc/math/vector/Vector.hpp]: There is an issue with this change here and in other lines: the `dim3` macro is still available and is substituted first.

**Comment by psychocoderHPC** [comment]
I will wait until we can switch to cupla where https://github.com/ComputationalRadiationPhysics/cupla/pull/161 is merged, this will simplify the cleanup.

File: include/pmacc/cuSTL/container/copier/Memcopy.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/pmacc/cuSTL/container/copier/Memcopy.hpp]: changed to `CUPLA_CHECK`

File: include/pmacc/math/vector/Vector.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/pmacc/math/vector/Vector.hpp]: solved by using `cupla.hpp` instead of the redefinition macros

**Comment by psychocoderHPC** [comment]
should be ready to merge

**Comment by psychocoderHPC** [comment]
cc-ing: @KseniaBastrakova After this PR is merged you need to check if your current refactored merge plugin is compiling. It could be you must change `atomicAdd(...` to `cupla::atomicAdd(acc,..)` and `__synchrthread()` ot `cupla::__synchrthread(acc)`

**Pull request merged**